Document using SOCKS5 to access Kubernetes API (#31527)
kubernetes/website: content/en/docs/tasks/extend-kubernetes/socks5-proxy-access-api.md

---
title: Use a SOCKS5 Proxy to Access the Kubernetes API
content_type: task
weight: 42
min-kubernetes-server-version: v1.24.0
---

<!-- overview -->
This page shows how to use SOCKS5 proxying to access the API of a remote Kubernetes cluster. This is useful when the cluster you want to access does not expose its API directly on the public internet.

## {{% heading "prerequisites" %}}

{{< include "task-tutorial-prereqs.md" >}} {{< version-check >}}

<!-- steps -->

## Overview

{{< note >}}
This example tunnels traffic using SSH, with the SSH client and server acting as a SOCKS proxy. You can instead use any other kind of [SOCKS5](https://en.wikipedia.org/wiki/SOCKS#SOCKS5) proxy.
{{</ note >}}

> The following diagram represents what we're going to achieve in this tutorial:

Reviewer: ```suggestion
Figure 1 represents what we're going to achieve in this tutorial.

* We've got a _client machine_ from where we're going to create requests to talk to the Kubernetes API
* The Kubernetes server/API is hosted on a _remote server_.
* We leverage SSH to create a secure SOCKS5 tunnel between the _local_ and the _remote server_ on top of which the HTTPS traffic between the client and the Kubernetes API will flow.

{{< mermaid >}}
graph LR;

  subgraph local[Local client machine]
  client([client])-- local <br> traffic .->  local_ssh[Local SSH <br> SOCKS5 proxy];
  end
  local_ssh[SSH <br>SOCKS5 <br> proxy]-- SSH Tunnel -->sshd
  
  subgraph remote[Remote server]
  sshd[SSH <br> server]-- local traffic -->service1;
  end
  client([client])-. proxied HTTPs traffic <br> going through the proxy .->service1[Kubernetes API];

  classDef plain fill:#ddd,stroke:#fff,stroke-width:4px,color:#000;
  classDef k8s fill:#326ce5,stroke:#fff,stroke-width:4px,color:#fff;
  classDef cluster fill:#fff,stroke:#bbb,stroke-width:2px,color:#326ce5;
  class ingress,service1,service2,pod1,pod2,pod3,pod4 k8s;
  class client plain;
  class cluster cluster;
{{</ mermaid >}}

> 

Reviewer: ```suggestion
Figure 1. SOCKS5 tutorial components

```

---
title: Use a SOCKS5 Proxy to Access the Kubernetes API
content_type: task
weight: 42

> min-kubernetes-server-version: v1.24.0

Reviewer 2: should this be `min-kubernetes-version: 1.24`? I'm not sure.

Author: you're right, corrected.

---

<!-- overview -->

> This page shows how to use SOCKS5 proxying to access the API of a remote Kubernetes cluster. This is useful when the cluster you want to access does not expose its API directly on the public internet.

Reviewer 2: ```suggestion
This page shows how to use a SOCKS5 proxy to access the API of a remote Kubernetes cluster.
This is useful when the cluster you want to access does not expose its API directly on the public internet.
```

## {{% heading "prerequisites" %}}

{{< include "task-tutorial-prereqs.md" >}} {{< version-check >}}

<!-- steps -->

## Overview

{{< note >}}

> This example tunnels traffic using SSH, with the SSH client and server acting as a SOCKS proxy. You can instead use any other kind of [SOCKS5](https://en.wikipedia.org/wiki/SOCKS#SOCKS5) proxy.

Reviewer 2: ```suggestion
This example tunnels traffic using SSH, with the SSH client and server acting as a SOCKS proxy. 
You can instead use any other kind of [SOCKS5](https://en.wikipedia.org/wiki/SOCKS#SOCKS5) proxies.
```

* We've got a _client machine_ from where we're going to create requests to talk to the Kubernetes API
* The Kubernetes server/API is hosted on a _remote server_.
* We leverage SSH to create a secure SOCKS5 tunnel between the _local_ and the _remote server_ on top of which the HTTPS traffic between the client and the Kubernetes API will flow.

{{< mermaid >}}
graph LR;

  subgraph local[Local client machine]
  client([client])-- local <br> traffic .->  local_ssh[Local SSH <br> SOCKS5 proxy];
  end
  local_ssh[SSH <br>SOCKS5 <br> proxy]-- SSH Tunnel -->sshd
  
  subgraph remote[Remote server]
  sshd[SSH <br> server]-- local traffic -->service1;
  end
  client([client])-. proxied HTTPs traffic <br> going through the proxy .->service1[Kubernetes API];

  classDef plain fill:#ddd,stroke:#fff,stroke-width:4px,color:#000;
  classDef k8s fill:#326ce5,stroke:#fff,stroke-width:4px,color:#fff;
  classDef cluster fill:#fff,stroke:#bbb,stroke-width:2px,color:#326ce5;
  class ingress,service1,service2,pod1,pod2,pod3,pod4 k8s;
  class client plain;
  class cluster cluster;
{{</ mermaid >}}
Figure 1. SOCKS5 tutorial components

## Using ssh to create a SOCKS5 proxy

> This command starts a SOCKS5 proxy between your _client machine_ and the _remote server_ where the Kubernetes API is listening:

Reviewer 2: Please wrap the long lines properly for ease of change tracking.
Also, please avoid adding italics when it is not necessary. Italics doesn't always render well for some language.

```suggestion
This command starts a SOCKS5 proxy between your client machine and the remote server
where the Kubernetes API is listening:
```

```shell
ssh -D 8080 -q -N username@kubernetes-jump-box.example
```

* `-D 8080`: opens a SOCKS proxy on local port :8080.
* `-q`: quiet mode. Causes most warning and diagnostic messages to be suppressed.
* `-N`: Do not execute a remote command. Useful for just forwarding ports.
* `username@kubernetes-jump-box.example`: the remote SSH server where the Kubernetes cluster is running.

## Client configuration

> To explore the Kubernetes API you'll first need to instruct your clients to send their queries through the SOCKS5 proxy we created earlier:

Reviewer 2: ```suggestion
To explore the Kubernetes API you'll first need to instruct your clients to send their queries through the SOCKS5 proxy we created earlier.
```

> for command-line tools, set the `https_proxy` environment variable in your and pass it to commands that you run.

Reviewer 2: ```suggestion
For command-line tools, set the `https_proxy` environment variable and pass it to commands that you run.
```

```shell
export https_proxy=socks5://localhost:8080
```

> When you set the `https_proxy` variable, tool such as `curl` route HTTPS traffic through the proxy that you configured. For this to work, the tool must support SOCKS5 proxying.

Reviewer 2: ```suggestion
When you set the `https_proxy` variable, tools such as `curl` route HTTPS traffic through the proxy
you configured. For this to work, the tool must support SOCKS5 proxying.
```

{{< note >}}

> In this example localhost will not be the localhost of the _client machine_ but the localhost of the _remote server_. It could also be replaced by a hostname that is known by the _remote server_ and hosted further in the remote infrastructure.

Reviewer 2: The second sentence doesn't make senses.

Author: corrected

---
title: Use a SOCKS5 Proxy to Access the Kubernetes API
content_type: task
weight: 42
min-kubernetes-server-version: v1.24
---

> 
> <!-- overview -->

Reviewer 3: I presume that this is instantly stable in v1.24? If so:
```suggestion
<!-- overview -->

{{< feature-state for_k8s_version="v1.24" state="stable" >}}

```

This page shows how to use a SOCKS5 proxy to access the API of a remote Kubernetes cluster.
This is useful when the cluster you want to access does not expose its API directly on the public internet.

## {{% heading "prerequisites" %}}

{{< include "task-tutorial-prereqs.md" >}} {{< version-check >}}

> 

Reviewer 3: ```suggestion
To follow this task you need an intermediate host (also known as a “jump box”). The task assumes that
the intermediate host is separate from your Kubernetes cluster.
You need SSH client software (the `ssh` tool), and an SSH server running on the intermediate host.
You must be able to log in to the SSH server on the intermediate host.

```

<!-- steps -->

> ## Overview

Reviewer 3: ```suggestion
## Task context
```

{{< note >}}
This example tunnels traffic using SSH, with the SSH client and server acting as a SOCKS proxy.
You can instead use any other kind of [SOCKS5](https://en.wikipedia.org/wiki/SOCKS#SOCKS5) proxies.
{{</ note >}}

> Figure 1 represents what we're going to achieve in this tutorial.

Reviewer 3: ```suggestion
Figure 1 represents what you're going to achieve in this task.
```
If this is a tutorial, it should be in the tutorial section. I think this _is_ a task page.

* We've got a client machine from where we're going to create requests to talk to the Kubernetes API
* The Kubernetes server/API is hosted on a remote server.
* We leverage SSH to create a secure SOCKS5 tunnel between the local and the remote server on top of

>   which the HTTPS traffic between the client and the Kubernetes API will flow.

Reviewer 3: ```
* You have a client compute from where you're going to create requests to talk to the Kubernetes API
* The Kubernetes server/API is hosted on a remote server.
* You will use SSH client and server software to create a secure SOCKS5 tunnel between the local and
  the remote server. The HTTPS traffic between the client and the Kubernetes API will flow over the SOCKS5
  tunnel, which is itself tunnelled over SSH.
```

{{< mermaid >}}
graph LR;

  subgraph local[Local client machine]
  client([client])-- local <br> traffic .->  local_ssh[Local SSH <br> SOCKS5 proxy];
  end
  local_ssh[SSH <br>SOCKS5 <br> proxy]-- SSH Tunnel -->sshd
  
  subgraph remote[Remote server]
  sshd[SSH <br> server]-- local traffic -->service1;
  end
  client([client])-. proxied HTTPs traffic <br> going through the proxy .->service1[Kubernetes API];

  classDef plain fill:#ddd,stroke:#fff,stroke-width:4px,color:#000;
  classDef k8s fill:#326ce5,stroke:#fff,stroke-width:4px,color:#fff;
  classDef cluster fill:#fff,stroke:#bbb,stroke-width:2px,color:#326ce5;
  class ingress,service1,service2,pod1,pod2,pod3,pod4 k8s;
  class client plain;
  class cluster cluster;
{{</ mermaid >}}
Figure 1. SOCKS5 tutorial components

## Using ssh to create a SOCKS5 proxy

This command starts a SOCKS5 proxy between your client machine and the remote server
where the Kubernetes API is listening:

```shell

> ssh -D 8080 -q -N username@kubernetes-jump-box.example

Reviewer 3: ```suggestion
# The SSH tunnel continues running in the background after you run this
ssh -D 8080 -q -N username@kubernetes-jump-box.example
```

```

> * `-D 8080`: opens a SOCKS proxy on local port :8080.

Reviewer 3: Port 8080 is very often used for HTTP traffic (for example, during software development). I would pick another port. How about port 1080, the SOCKS port?

* `-q`: quiet mode. Causes most warning and diagnostic messages to be suppressed.
* `-N`: Do not execute a remote command. Useful for just forwarding ports.
* `username@kubernetes-jump-box.example`: the remote SSH server where the Kubernetes cluster is running.

## Client configuration

To explore the Kubernetes API you'll first need to instruct your clients to send their queries through
the SOCKS5 proxy we created earlier.

For command-line tools, set the `https_proxy` environment variable and pass it to commands that you run.

```shell

> export https_proxy=socks5://localhost:8080

Reviewer 3: ```suggestion
export https_proxy=socks5h://localhost:8080
```
This forces `curl` to have the remote side do the DNS resolution.

```

When you set the `https_proxy` variable, tools such as `curl` route HTTPS traffic through the proxy
you configured. For this to work, the tool must support SOCKS5 proxying.

{{< note >}}

> In this example localhost will not be the localhost of the client machine but the localhost of
> the remote server.
> It could also be replaced by a hostname that is known by the remote server.

Reviewer 3: ```suggestion
In the URL https://localhost/api, `localhost` does not refer to your local client computer.
Instead, it refers to the the endpoint that the intermediate host (“jump box”) knows as `localhost`.
The `curl` tool sends the hostname from the HTTPS URL over SOCKS, and the intermediate
host resolves that locally (to an address that belongs to its loopback interface).
```

Reviewer 3: but see comments below

{{</ note >}}

```shell
curl -k -v https://localhost/api
```

To use the official Kubernetes client `kubectl` with a proxy, set the `proxy-url` element
for the relevant `cluster` entry within  your `~/.kube/config` file. For example:

```yaml
apiVersion: v1
clusters:
- cluster:
    certificate-authority-data: LRMEMMW2 # shortened for readability 
    server: https://localhost            # the "Kubernetes API" in the diagram above

>     proxy-url: socks5://localhost:8080   # the "SSH SOCKS5 proxy" in the diagram above

Reviewer 3: I've added a note about where the DNS resolution happens, making the environment variable match what's documented. _However_, I don't know whether `kubectl` and friends resolve hostnames locally or remotely when using SOCKS v5. So this might be wrong and instead we need to change the meaning of the note about name resolution.

If kubectl does default to having name resolution performed by the remote peer, we could add a comment about why the URL scheme here is not `socks5h` (I presume the answer would then be: DNS resolution over socks is the default).

Author: I'm 99% sure that resolution is done by the remote peer but I'll need to validate that in a lab. I'll keep this comment open for now.

Author: From what I've [read](https://github.com/hashicorp/terraform/issues/17754#issuecomment-386411167) and tested kubectl supports remote DNS resolution through socks5 proxies without `socks5h://` or `socks5://` in front of the proxy URL.

* The Kubernetes server/API is hosted on a remote server.
* You will use SSH client and server software to create a secure SOCKS5 tunnel between the local and
  the remote server. The HTTPS traffic between the client and the Kubernetes API will flow over the SOCKS5
  tunnel, which is itself tunnelled over SSH.

{{< mermaid >}}
graph LR;

  subgraph local[Local client machine]
  client([client])-- local <br> traffic .->  local_ssh[Local SSH <br> SOCKS5 proxy];
  end
  local_ssh[SSH <br>SOCKS5 <br> proxy]-- SSH Tunnel -->sshd
  
  subgraph remote[Remote server]
  sshd[SSH <br> server]-- local traffic -->service1;
  end
  client([client])-. proxied HTTPs traffic <br> going through the proxy .->service1[Kubernetes API];

  classDef plain fill:#ddd,stroke:#fff,stroke-width:4px,color:#000;
  classDef k8s fill:#326ce5,stroke:#fff,stroke-width:4px,color:#fff;
  classDef cluster fill:#fff,stroke:#bbb,stroke-width:2px,color:#326ce5;
  class ingress,service1,service2,pod1,pod2,pod3,pod4 k8s;
  class client plain;
  class cluster cluster;
{{</ mermaid >}}
Figure 1. SOCKS5 tutorial components

## Using ssh to create a SOCKS5 proxy

> This command starts a SOCKS5 proxy between your client machine and the remote server
> where the Kubernetes API is listening:

Reviewer 3: ```suggestion
This command starts a SOCKS5 proxy between your client machine and the intermediate
host. The SOCKS5 proxy lets you connect via that immediate host to your cluster's API
server (provided that the traffic is permitted).
```

curl -k -v https://localhost/api
```

To use the official Kubernetes client `kubectl` with a proxy, set the `proxy-url` element
for the relevant `cluster` entry within  your `~/.kube/config` file. For example:

```yaml
apiVersion: v1
clusters:
- cluster:
    certificate-authority-data: LRMEMMW2 # shortened for readability 
    server: https://localhost            # the "Kubernetes API" in the diagram above
    proxy-url: socks5://localhost:1080   # the "SSH SOCKS5 proxy" in the diagram above
  name: default
contexts:
- context:
    cluster: default
    user: default
  name: default
current-context: default
kind: Config
preferences: {}
users:
- name: default
  user:
    client-certificate-data: LS0tLS1CR== # shortened for readability
    client-key-data: LS0tLS1CRUdJT=      # shortened for readability
```

If the tunnel is operating and you use `kubectl` with a context that uses this cluster, you can interact with your cluster through that proxy. For example:

```shell
kubectl get pods
```

```console
NAMESPACE     NAME                                     READY   STATUS      RESTARTS   AGE
kube-system   coredns-85cb69466-klwq8                  1/1     Running     0          5m46s

> ```

Reviewer 3: You could optionally add a “Clean up” heading that explains how to stop the existing `ssh` process. Possibly also some further reading with links to OpenSSH docs?

---
title: Use a SOCKS5 Proxy to Access the Kubernetes API
content_type: task
weight: 42
min-kubernetes-server-version: v1.24
---
<!-- overview -->

{{< feature-state for_k8s_version="v1.24" state="stable" >}}

This page shows how to use a SOCKS5 proxy to access the API of a remote Kubernetes cluster.
This is useful when the cluster you want to access does not expose its API directly on the public internet.

## {{% heading "prerequisites" %}}

{{< include "task-tutorial-prereqs.md" >}} {{< version-check >}}

> You need SSH client software (the `ssh` tool), and an SSH server running on the remote server.

Reviewer 4: Referring to it as a service would help clarify, I believe.
```suggestion
You need SSH client software (the `ssh` tool), and an SSH service running on the remote server.
```

OR 

```suggestion
You need SSH client software (the `ssh` tool), and an SSH server running on the remote host.
```

> You must be able to log in to the SSH server on the remote server.

Reviewer 4: Similar to the above.

```suggestion
You must be able to log in to the SSH server on the remote host.
```

OR 

```suggestion
You must be able to log in to the SSH service on the remote server.
```

<!-- steps -->

## Task context

{{< note >}}
This example tunnels traffic using SSH, with the SSH client and server acting as a SOCKS proxy.
You can instead use any other kind of [SOCKS5](https://en.wikipedia.org/wiki/SOCKS#SOCKS5) proxies.
{{</ note >}}

Figure 1 represents what you're going to achieve in this task.

> * You have a client computer from where you're going to create requests to talk to the Kubernetes API

Reviewer 4: Do we want to mention here that the client computer will be referred to as local in the diagram/steps ahead? I understand we explicitly speak about how localhost is not the local client computer somewhere below. But I think it'd help if we state this explicitly for step #3.

```suggestion
* You have a client computer, referred to as local in the steps ahead, from where you're going to create requests to talk to the Kubernetes API.
```

server (provided that the traffic is permitted).

```shell
# The SSH tunnel continues running in the foreground after you run this
ssh -D 1080 -q -N username@kubernetes-jump-box.example
```

* `-D 1080`: opens a SOCKS proxy on local port :1080.
* `-q`: quiet mode. Causes most warning and diagnostic messages to be suppressed.
* `-N`: Do not execute a remote command. Useful for just forwarding ports.
* `username@kubernetes-jump-box.example`: the remote SSH server where the Kubernetes cluster is running.

## Client configuration

To explore the Kubernetes API you'll first need to instruct your clients to send their queries through
the SOCKS5 proxy we created earlier.

For command-line tools, set the `https_proxy` environment variable and pass it to commands that you run.

```shell
export https_proxy=socks5h://localhost:1080
```

When you set the `https_proxy` variable, tools such as `curl` route HTTPS traffic through the proxy
you configured. For this to work, the tool must support SOCKS5 proxying.

{{< note >}}
In the URL https://localhost/api, `localhost` does not refer to your local client computer.

> Instead, it refers to the the endpoint that the intermediate host (“jump box”) knows as `localhost`.

Reviewer 4: ```suggestion
Instead, it refers to the endpoint that the intermediate host (“jump box”) knows as `localhost`.
```

The `curl` tool sends the hostname from the HTTPS URL over SOCKS, and the intermediate

> host resolves that locally (to an address that belongs to its loopback interface).

Reviewer 5: at the beginning you say

> To follow this task you need an intermediate host (also known as a “jump box”). The task assumes that
the intermediate host is separate from your Kubernetes cluster.

but here it assumes the apiserver is running on the jump box, since you connect to localhost of the jumpbox, or am I getting it wrong?

Author: @aojea ohhh thank you so much for catching that! You're 100% right. That's what I meant initially, to keep it simple the ssh-server **is** running on the same machine as the kubernetes server.
It doesn't mean that it can't run in a dedicated remote machine (via a jump-box) but this is outside the scope of this tutorial.

I removed this notion of "jump-box" that was introduced during the review process and reworked slightly the phrasing on the document

ptal :)

{{< mermaid >}}
graph LR;

  subgraph local[Local client machine]
  client([client])-- Local <br> traffic .->  local_ssh[Local SSH <br> SOCKS5 proxy];
  end
  local_ssh[SSH <br>SOCKS5 <br> proxy]-- SSH Tunnel -->SSHd
  
  subgraph remote[Remote server]
  SSHd[SSHd <br> server]-- Local traffic -->service1;
  end
  client([client])-. proxied HTTPs traffic <br> going through the proxy .->service1[Kubernetes API];

  classDef plain fill:#ddd,stroke:#fff,stroke-width:4px,color:#000;
  classDef k8s fill:#326ce5,stroke:#fff,stroke-width:4px,color:#fff;
  classDef cluster fill:#fff,stroke:#bbb,stroke-width:2px,color:#326ce5;
  class ingress,service1,service2,pod1,pod2,pod3,pod4 k8s;
  class client plain;
  class cluster cluster;
{{</ mermaid >}}

## Using ssh to create a SOCKS5 proxy

> This command starts a SOCKS5 proxy between your client machine and the target where the Kubernetes API is listening:
> 
>     $ssh -D 8080 -q -N  username@kubernetes_server
> 
> * -D 8080: opens a SOCKS proxy on local port :8080.
> * -q: quiet mode. Causes most warning and diagnostic messages to be suppressed.
> * -N: Do not execute a remote command. Useful for just forwarding ports.
> * username@kubernetes_server: the remote SSH server where the Kubernetes cluster is running.

Reviewer 6: https://kubernetes.io/docs/contribute/style/style-guide/#code-snippet-formatting

- All `$`s at the beginning of commands should be excluded.
- In Style guide, explicit code block notations rather than implicit (using indents) one is used. 😊

And when you apply this suggestion,
please remove the word `shell` in L47! 😊

```suggestion
This command starts a SOCKS5 proxy between your client machine and the target where the Kubernetes API is listening:

```shell
ssh -D 8080 -q -N  username@kubernetes_server
```shell

* `-D 8080`: opens a SOCKS proxy on local port `8080`.
* `-q`: quiet mode. Causes most warning and diagnostic messages to be suppressed.
* `-N`: Do not execute a remote command. Useful for just forwarding ports.
* `username@kubernetes_server`: the remote SSH server where the Kubernetes cluster is running.
```

Author: corrected!

## Client configuration

To explore the Kubernetes API you'll first need to instruct your clients to send their queries through the SOCKS5 proxy we created earlier:

> for command-line tools we can export the following ENV variable:
> 
>     $export HTTPS_PROXY=socks5://localhost:8080
> 
> Commands like `cURL` will then route their traffic through the proxy:
> 
>     $curl -k -v https://localhost/api
> 
> In this example localhost will not be the localhost of the client machine but the localhost of the Kubernetes server.
> 
> To use the official Kubernetes client `kubectl` with a proxy you'll need to set it in your `~/.kube/config` file:
> 
>     - cluster:
>           certificate-authority-data: xxx
>           server: https://localhost:6443
>           proxy-url: socks5://localhost:8080
> 
> Then you should be able to query the cluster through the proxy:
> 
>     $kubectl get pods
>     NAMESPACE     NAME                                     READY   STATUS      RESTARTS   AGE
>     kube-system   coredns-85cb69466-klwq8                  1/1     Running     0          5m46s

Reviewer 6: For anyone: It will be great to confirm that `~/.kube/config` file is using YAML format. I am not 100% sure! 😅

```suggestion
for command-line tools we can export the following ENV variable:

```shell
export HTTPS_PROXY=socks5://localhost:8080
```shell

Commands like `cURL` will then route their traffic through the proxy:

```shell
curl -k -v https://localhost/api
```shell

In this example localhost will not be the localhost of the client machine but the localhost of the Kubernetes server.

To use the official Kubernetes client `kubectl` with a proxy you'll need to set it in your `~/.kube/config` file:

```yaml
- cluster:
  certificate-authority-data: xxx
  server: https://localhost:6443
  proxy-url: socks5://localhost:8080
```yaml

Then you should be able to query the cluster through the proxy:

```shell
kubectl get pods
```shell

```console
NAMESPACE     NAME                                     READY   STATUS      RESTARTS   AGE
kube-system   coredns-85cb69466-klwq8                  1/1     Running     0          5m46s
```console
```

Author: I confirm it's yaml

Author: Done!

---
title: Use a SOCKS5 Proxy to Access the Kubernetes API
content_type: task
weight: 40
---

<!-- overview -->
This page shows how to use an SOCKS5 proxy to access a remote Kubernetes API.

## {{% heading "prerequisites" %}}

{{< include "task-tutorial-prereqs.md" >}} {{< version-check >}}

<!-- steps -->

## Setup big-picture

>   {{< note >}}
>   In this example we're using SSH as the tunnel mechanism but it can be replaced by any other kind of SOCKS5 proxy.
>   {{</ note >}}

Reviewer 6: Let's remove non-essential indents. 😊

```suggestion
{{< note >}}
In this example we're using SSH as the tunnel mechanism but it can be replaced by any other kind of SOCKS5 proxy.
{{</ note >}}
```

Author: done thanks!

```shell
ssh -D 8080 -q -N  username@kubernetes_server
```

* `-D 8080`: opens a SOCKS proxy on local port :8080.
* `-q`: quiet mode. Causes most warning and diagnostic messages to be suppressed.
* `-N`: Do not execute a remote command. Useful for just forwarding ports.
* `username@kubernetes_server`: the remote SSH server where the Kubernetes cluster is running.

## Client configuration

To explore the Kubernetes API you'll first need to instruct your clients to send their queries through the SOCKS5 proxy we created earlier:

for command-line tools we can export the following ENV variable:

```shell
export HTTPS_PROXY=socks5://localhost:8080
```

Commands like `cURL` will then route their traffic through the proxy:

{{< note >}}
In this example localhost will not be the localhost of the `client machine` but the localhost of the `remote server`. It could also be replaced by a hostname that is known by the `remote server` and hosted further in the remote infrastructure.
{{</ note >}}

```shell
curl -k -v https://localhost/api
```

To use the official Kubernetes client `kubectl` with a proxy you'll need to set it in your `~/.kube/config` file:

```yaml

> - cluster:
>       certificate-authority-data: xxx
>       server: https://localhost
>       proxy-url: socks5://localhost:8080

Reviewer 6: For YAML, people usually indent with 2 spaces. 😊

And it will be great if you test this example is working with no problem! 😊

```suggestion
- cluster:
  certificate-authority-data: xxx
  server: https://localhost
  proxy-url: socks5://localhost:8080
```

Author: corrected + added a note that mention that the example in the yaml snippet must be added inside the config (this is not the whole config file).

```

Then you should be able to query the remote cluster through the proxy:

> ```console
> kubectl get pods
> NAMESPACE     NAME                                     READY   STATUS      RESTARTS   AGE
> kube-system   coredns-85cb69466-klwq8                  1/1     Running     0          5m46s

Reviewer 6: We should [separate commands from output](https://kubernetes.io/docs/contribute/style/style-guide/#separate-commands-from-output)!

```suggestion
```shell
kubectl get pods
```shell

```console
NAMESPACE     NAME                                     READY   STATUS      RESTARTS   AGE
kube-system   coredns-85cb69466-klwq8                  1/1     Running     0          5m46s
```

Author: make sense, corrected!

---
title: Use a SOCKS5 Proxy to Access the Kubernetes API
content_type: task

> weight: 40

Reviewer 3: ```suggestion
weight: 42
```
I'd order this after the `kubectl proxy` task.

---

<!-- overview -->
This page shows how to use an SOCKS5 proxy to access a remote Kubernetes API.

## {{% heading "prerequisites" %}}

> {{< include "task-tutorial-prereqs.md" >}} {{< version-check >}}

Author: Thanks added the version requirement (version has not been released yet).

<!-- steps -->

## Setup big-picture

{{< note >}}
In this example we're using SSH as the tunnel mechanism but it can be replaced by any other kind of SOCKS5 proxy.
{{</ note >}}

The following diagram represents what we're going to achieve in this tutorial:

* We've got a `client machine` from where we're going to create requests to talk to the Kubernetes API
* The Kubernetes server/API is hosted on a `remote server`.

> * We leverage SSH to create a secure SOCKS5 tunnel between the `local` and the `remote server` on top of which the HTTPS traffic between the client and the Kubernetes API will flow.

Reviewer 3: Here I'd use underscores (italics) not backticks (which are for inline code).

{{< mermaid >}}
graph LR;

  subgraph local[Local client machine]
  client([client])-- local <br> traffic .->  local_ssh[Local SSH <br> SOCKS5 proxy];
  end

>   local_ssh[SSH <br>SOCKS5 <br> proxy]-- SSH Tunnel -->SSHd
>   
>   subgraph remote[Remote server]
>   SSHd[SSHd <br> server]-- local traffic -->service1;

Reviewer 3: ```suggestion
  local_ssh[SSH <br>SOCKS5 <br> proxy]-- SSH Tunnel -->sshd

  subgraph remote[Remote server]
  sshd[SSH <br> server]-- local traffic -->service1;
```

The capitalization “SSHd” is not commonly used.

  end
  client([client])-. proxied HTTPs traffic <br> going through the proxy .->service1[Kubernetes API];

  classDef plain fill:#ddd,stroke:#fff,stroke-width:4px,color:#000;
  classDef k8s fill:#326ce5,stroke:#fff,stroke-width:4px,color:#fff;
  classDef cluster fill:#fff,stroke:#bbb,stroke-width:2px,color:#326ce5;
  class ingress,service1,service2,pod1,pod2,pod3,pod4 k8s;
  class client plain;
  class cluster cluster;
{{</ mermaid >}}

## Using ssh to create a SOCKS5 proxy

> This command starts a SOCKS5 proxy between your `client machine` and the `remote server` where the Kubernetes API is listening:

Reviewer 3: Here I'd use underscores (italics) not backticks (which are for inline code).

```shell
ssh -D 8080 -q -N  username@kubernetes_server
```

* `-D 8080`: opens a SOCKS proxy on local port :8080.
* `-q`: quiet mode. Causes most warning and diagnostic messages to be suppressed.
* `-N`: Do not execute a remote command. Useful for just forwarding ports.
* `username@kubernetes_server`: the remote SSH server where the Kubernetes cluster is running.

## Client configuration

To explore the Kubernetes API you'll first need to instruct your clients to send their queries through the SOCKS5 proxy we created earlier:

> for command-line tools we can export the following ENV variable:

Reviewer 3: I'd write:
```suggestion
for command-line tools, set the `https_proxy` environment variable in your and pass it to commands that you run
```

See https://superuser.com/questions/876100/https-proxy-vs-https-proxy for more details about case of this variable. `HTTP_PROXY` carries a security risk and so the `*_proxy` variables are usually lower case.

```shell
export HTTPS_PROXY=socks5://localhost:8080
```

> Commands like `cURL` will then route their traffic through the proxy:

Reviewer 3: ```suggestion
When you set the `https_proxy` variable, tool such as `curl` route HTTPS traffic
through the proxy that you configured. For this to work, the tool must support SOCKS5
proxying.
```

{{< note >}}
In this example localhost will not be the localhost of the `client machine` but the localhost of the `remote server`. It could also be replaced by a hostname that is known by the `remote server` and hosted further in the remote infrastructure.
{{</ note >}}

```shell
curl -k -v https://localhost/api
```

To use the official Kubernetes client `kubectl` with a proxy you'll need to add the `proxy-url` element inside the `cluster` entry of  your `~/.kube/config` file, like:

```yaml
- cluster:
    proxy-url: socks5://localhost:8080
    server: https://localhost
```

> Then you should be able to query the remote cluster through the proxy:

Reviewer 3: ```suggestion
If the tunnel is operating and you use `kubectl` with a context that uses this cluster, you can interact with your
cluster through that proxy. For example:
```
